### PR TITLE
ENH: `stats.ttest_ind`: add array API support

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-20544

#### What does this implement/fix?
Adds array API support to `scipy.stats.ttest_ind`.

#### Additional information
Trimmed t-test not yet supported. Let's leave that for a separate PR. (I don't personally plan to do it.)

### CODE REVIEW

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7540]: I forgot to add a test for the new errors. If there's nothing else, let's add it in the follow-up that tackles the trimmed t-test.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7515]: Yes, I changed the type of this error. It should have been a `NotImplementedError` originally.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7519]: I moved this error into the `_permutation_ttest` function.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7553]: Suggestion: review with "Hide whitespace". I added an early return here to dedent the important part of the function.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7535]: I think the reason for this unusual form of `np.errstate` was to only apply it when `equal_var` is `True`. I don't think this is really needed, so I've simplified it.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7702]: This can be changed back if desired, but it's not worth re-running CI.
```suggestion
    return (avg_a-avg_b)/denom
```
The reason is that I was going to do `res[()] if res.ndim == 0 else res` but it didn't seem to be needed.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7748]: Moved from inside `ttest_ind`.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5061]: Moved to separate test only for NumPy

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5068]: Just removed this. Not a strong test. Not worth the headache.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5111]: Otherwise missing since I've moved `nan_policy` tests to a separate NumPy-only function.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5184]:
```suggestion
    # test scalars
```
Copy-pasted out of `test_ttest_ind`.

**Review by mdhaber — COMMENTED** [review]
Some comments to aid the reviewer.

I would be willing to review a PR that refactors the sprawling tests. For instance, we can can come close to halving the length of some tests by reusing code when testing `ttest_ind` and `ttest_ind_from_stats` rather than copy-pasting.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5114]:
```suggestion
```
doh

**Comment by mdhaber** [comment]
This is ready for review but I'd prefer to merge this after gh-20694 to keep the merge conflicts there more manageable.

File: scipy/stats/_stats_py.py
Comment (j-bowhay) [reply scipy/stats/_stats_py.py:7002]: Would be nice to do this at the same this as resolving merge conflicts

**Comment by mdhaber** [comment]
Failures seem unrelated. One is a datasets timeout, and the other is a failure elsewhere that I'll investigate shortly.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [reply scipy/stats/_stats_py.py:6977]: Done!

File: scipy/stats/tests/test_stats.py
Comment (lucascolley) [inline scipy/stats/tests/test_stats.py:5699]: The top of this file has `skip_xp_backends = pytest.mark.skip_xp_backends`, if you'd like to reduce verbosity a tad

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:5699]: Ok. 
I think I saw a PR in which you tried to eliminate the need for `usefixture` line everywhere. Did that not work out?

Comment (lucascolley) [reply scipy/stats/tests/test_stats.py:5699]: hmm, there is gh-20368, but that isn't quite what it was trying to do - `usefixtures` is still needed wherever you are trying to skip. In any case, I couldn't get that PR working.

Comment (lucascolley) [reply scipy/stats/tests/test_stats.py:5699]: at the minute, you have to wait until every test in a test file is converted, then the verbosity can be reduced greatly with `pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends")]`

Comment (lucascolley) [reply scipy/stats/tests/test_stats.py:5699]: In this sense, there would be benefits to having smaller test files than `test_stats.py`. I suppose keeping the commit history is more important, though.

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:5699]: Hah yeah, I'd love to have smaller files. I did find a way of splitting the files while preserving the history. But that's probably better done when things are a bit more stable.

**Comment by lucascolley** [comment]
reason for changing the PR title: https://github.com/scipy/scipy/issues/20739#issuecomment-2119343758

File: scipy/stats/tests/test_stats.py
Comment (j-bowhay) [inline scipy/stats/tests/test_stats.py:5239]: optional but nice to modernise at the same time
```suggestion
    assert np.isnan(t)
    assert np.isnan(p)
```

File: scipy/stats/tests/test_stats.py
Comment (j-bowhay) [inline scipy/stats/tests/test_stats.py:6064]: Can you add a comment on why pytorch requires a lower tolerance? I am guessing because `x` is an integral type so in pytorch is cast to float32?

**Review by j-bowhay — COMMENTED** [review]
Some minor queries for my own understanding but looks in good shape

File: scipy/stats/_stats_py.py
Comment (mdhaber) [reply scipy/stats/_stats_py.py:6532]: `_unequal_var_ttest_denom` is called from a few functions, and I think `v1` and `v2` are guaranteed to be arrays, but `n1` and `n2` are often determined from `shape`, so they aren't arrays unless we explicitly convert them. We could explicilty convert them... But in that case, we could also require that `_unequal_var_ttest_denom` is passed `xp` directly, so we would never call `array_namespace`. LMK what you'd prefer.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:5239]: Please go ahead and add this in the `ttest_rel` PR.

File: scipy/stats/_stats_py.py
Comment (j-bowhay) [reply scipy/stats/_stats_py.py:6532]: Happy to leave as is following discussion

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:6064]: `1e-10` was used as the tolerance in the original test, so I wanted to preserve, but `torch` is tested with its default dtype of `float32`, so it can't possibly meet that tolerance.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [reply scipy/stats/_stats_py.py:6957]: I think I'm waiting for consensus one what should be done in this situation before factoring out this code. Ultimately it could go in the `_axis_nan_policy` decorator or something.

**Review by j-bowhay — APPROVED** [review]
Thanks for clarifications

**Pull request merged**
